extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java

```diff
@@ -158,174 +131,73 @@ public List<WritableFrameChannel> outputChannels()
-        // pending flush to the output channel.
-        if (!rowsToProcess.isEmpty()) {
-          lastPartitionIndex = rowsToProcess.size() - 1;
-          processRowsUpToLastPartition();
-          return ReturnOrAwait.runAgain();
-        }
-        return ReturnOrAwait.returnObject(Unit.instance());
-      } else {
-        return ReturnOrAwait.runAgain();
-      }
-    }
-
-    while (!frameCursor.isDone()) {
-      final ResultRow currentRow = rowSupplierFromFrameCursor.get();
-      if (outputRow == null) {
-        outputRow = currentRow;
-        rowsToProcess.add(currentRow);
-      } else if (comparePartitionKeys(outputRow, currentRow, partitionColumnNames)) {
-        // Add current row to the same batch of rows for processing.
-        rowsToProcess.add(currentRow);
-      } else {
-        lastPartitionIndex = rowsToProcess.size() - 1;
-        outputRow = currentRow.copy();
-        rowsToProcess.add(currentRow);
+      return ReturnOrAwait.runAgain();
+    } else if (inputChannel.isFinished()) {
+      if (rowId.get() == 0) {
+        runAllOpsOnBatch();
       }
-      frameCursor.advance();
 
-      if (rowsToProcess.size() > maxRowsMaterialized) {
-        // We don't want to materialize more than maxRowsMaterialized rows at any point in time, so process the pending batch.
-        processRowsUpToLastPartition();
-        ensureMaxRowsInAWindowConstraint(rowsToProcess.size());
+      // If there are still rows pending after operations, run again.
+      if (frameHasRowsPendingFlush()) {
         return ReturnOrAwait.runAgain();
       }
+      return ReturnOrAwait.returnObject(Unit.instance());
+    } else {
+      return ReturnOrAwait.awaitAll(inputChannels().size());
     }
-    return ReturnOrAwait.runAgain();
   }
 
-  /**
-   * @param listOfRacs Concat this list of {@link RowsAndColumns} to a {@link ConcatRowsAndColumns} to use as a single input for the operators to be run
-   */
-  private void runAllOpsOnMultipleRac(ArrayList<RowsAndColumns> listOfRacs)
+  private void initialiseOperator()
   {
-    Operator op = new Operator()
+    op = new Operator()
     {
       @Nullable
       @Override
       public Closeable goOrContinue(Closeable continuationObject, Receiver receiver)
       {
-        RowsAndColumns rac = new ConcatRowsAndColumns(listOfRacs);
-        ensureMaxRowsInAWindowConstraint(rac.numRows());
+        RowsAndColumns rac = frameRowsAndColsBuilder.build();
+        ensureMaxRowsInAWindowConstraint(rac.numRows(), maxRowsMaterialized);
         receiver.push(rac);
-        receiver.completed();
-        return null;
+
+        if (inputChannel.isFinished()) {
+          // Only call completed() when the input channel is finished.
+          receiver.completed();
+          return null; // Signal that the operator has completed its work
+        }
+
+        // Return a non-null continuation object to indicate that we want to continue processing.
```

Comment:
I don't think this comment and  the returned  `Closeable`  lives up to the contracts associated with the implemented method...but that didn't worked before this change either.....

Reply:
Since we have to re-use the same operator, I needed a way to keep it running..